token.go

```diff
@@ -0,0 +1,215 @@
+	t.doRefresh <- true
+}
+
+// will block until valid
+func (t *tokenMan) getToken() *oauthToken {
+	getTokenLock.Lock()
+	defer getTokenLock.Unlock()
+
+	if t.token.isValid() {
+		log.Debugf("returning existing token: %v", t.token)
+		return t.token
+	}
+
+	t.retrieveNewToken()
+	return t.token
+}
+
+func (t *tokenMan) close() {
+	log.Debug("close token manager")
+	close(t.doRefresh)
+}
+
+// don't call externally. will block until success.
+func (t *tokenMan) retrieveNewToken() {
+
+	log.Debug("Getting OAuth token...")
+	uriString := config.GetString(configProxyServerBaseURI)
+	uri, err := url.Parse(uriString)
+	if err != nil {
+		log.Panicf("unable to parse uri config '%s' value: '%s': %v", configProxyServerBaseURI, uriString, err)
+	}
+	uri.Path = path.Join(uri.Path, "/accesstoken")
+
+	retryIn := 5 * time.Millisecond
+	maxBackOff := maxBackoffTimeout
+	backOffFunc := createBackOff(retryIn, maxBackOff)
+	first := true
+
+	for {
+		if first {
+			first = false
+		} else {
+			backOffFunc()
+		}
+
+		form := url.Values{}
+		form.Set("grant_type", "client_credentials")
+		form.Add("client_id", config.GetString(configConsumerKey))
+		form.Add("client_secret", config.GetString(configConsumerSecret))
+		req, err := http.NewRequest("POST", uri.String(), bytes.NewBufferString(form.Encode()))
+		req.Header.Set("Content-Type", "application/x-www-form-urlencoded; param=value")
+		req.Header.Set("display_name", apidInfo.InstanceName)
+		req.Header.Set("apid_instance_id", apidInfo.InstanceID)
+		req.Header.Set("apid_cluster_Id", apidInfo.ClusterID)
+		req.Header.Set("status", "ONLINE")
+		req.Header.Set("plugin_details", apidPluginDetails)
+
+		if newInstanceID {
+			req.Header.Set("created_at_apid", time.Now().Format(time.RFC3339))
+		} else {
+			req.Header.Set("updated_at_apid", time.Now().Format(time.RFC3339))
+		}
+
+		client := &http.Client{Timeout: httpTimeout}
+		resp, err := client.Do(req)
+		if err != nil {
+			log.Errorf("Unable to Connect to Edge Proxy Server: %v", err)
+			continue
+		}
+
+		body, err := ioutil.ReadAll(resp.Body)
+		resp.Body.Close()
+		if err != nil {
+			log.Errorf("Unable to read EdgeProxy Sever response: %v", err)
+			continue
+		}
+
+		if resp.StatusCode != 200 {
+			log.Errorf("Oauth Request Failed with Resp Code: %d. Body: %s", resp.StatusCode, string(body))
+			continue
+		}
+
+		var token oauthToken
+		err = json.Unmarshal(body, &token)
+		if err != nil {
+			log.Errorf("unable to unmarshal JSON response '%s': %v", string(body), err)
+			continue
+		}
+
+		if token.ExpiresIn >= 0 {
+			token.ExpiresAt = time.Now().Add(time.Duration(token.ExpiresIn) * time.Millisecond)
+		} else {
+			// no expiration, arbitrarily expire about a year from now
+			token.ExpiresAt = time.Now().Add(365 * 24 * time.Hour)
+		}
+
+		log.Debugf("Got new token: %#v", token)
+
+		if newInstanceID {
+			newInstanceID = false
+			updateApidInstanceInfo()
+		}
+
+		t.token = &token
+		config.Set(configBearerToken, token.AccessToken)
```

Comment:
The expectation for all the consumers of the plugin is to get from config, as set before?